pkg/services/ngalert/api/api_testing.go

```diff
@@ -0,0 +1,85 @@
+package api
+
+import (
+	"fmt"
+	"net/http"
+	"net/url"
+	"strconv"
+	"time"
+
+	apimodels "github.com/grafana/alerting-api/pkg/api"
+	"github.com/grafana/grafana/pkg/api/response"
+	"github.com/grafana/grafana/pkg/infra/log"
+	"github.com/grafana/grafana/pkg/models"
+	"github.com/grafana/grafana/pkg/services/datasources"
+	"github.com/grafana/grafana/pkg/setting"
+	"github.com/grafana/grafana/pkg/tsdb"
+	"github.com/grafana/grafana/pkg/util"
+)
+
+type TestingApiSrv struct {
+	*AlertingProxy
+	Cfg             *setting.Cfg
+	DataService     *tsdb.Service
+	DatasourceCache datasources.CacheService
+	log             log.Logger
+}
+
+func (srv TestingApiSrv) RouteTestReceiverConfig(c *models.ReqContext, body apimodels.ExtendedReceiver) response.Response {
+	srv.log.Info("RouteTestReceiverConfig: ", "body", body)
+	return response.JSON(http.StatusOK, util.DynMap{"message": "success"})
+}
+
+func (srv TestingApiSrv) RouteTestRuleConfig(c *models.ReqContext, body apimodels.TestRulePayload) response.Response {
+	recipient := c.Params("Recipient")
+	if recipient == apimodels.GrafanaBackend.String() {
+		if body.Type() != apimodels.GrafanaBackend || body.GrafanaManagedCondition == nil {
+			return response.Error(http.StatusBadRequest, "unexpected payload", nil)
+		}
+		return conditionEval(c, *body.GrafanaManagedCondition, srv.DatasourceCache, srv.DataService, srv.Cfg)
+	}
+
+	if body.Type() != apimodels.LoTexRulerBackend {
+		return response.Error(http.StatusBadRequest, "unexpected payload", nil)
+	}
+
+	var path string
+	if datasourceID, err := strconv.ParseInt(recipient, 10, 64); err == nil {
+		ds, err := srv.DatasourceCache.GetDatasource(datasourceID, c.SignedInUser, c.SkipCache)
+		if err != nil {
+			return response.Error(http.StatusInternalServerError, "failed to get datasource", err)
+		}
+
+		switch ds.Type {
+		case "loki":
+			path = "loki/api/v1/query_range"
```

Comment:
Rules are evaluated at one particular moment in time (now) and actually would use the `/query` instead of `/query_range` under the hood. I can see it being helpful, however, to have the history that `/query_range` provides in a test endpoint. Is this what you're intending to do?

Reply:
1. I had noticed the both `/query` and `/query_range` are called but I thought that for our case they were interchangeable so I just picked one. So no, it was not intentional.
2. We could pass a specific time as a query parameter (same as it happens for the grafana rules) so that we could test the query for different times. WDYT?